dialogue-core/src/main/java/com/palantir/dialogue/core/CostBasedRouting.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.codahale.metrics.ExponentiallyDecayingReservoir;
+import com.codahale.metrics.Reservoir;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.google.common.base.Stopwatch;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.logsafe.Preconditions;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.function.LongConsumer;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+final class CostBasedRouting implements LimitedChannel {
+    private final ImmutableList<LimitedChannel> channels;
+    // Each Reservoir stores the perceived cost of making a request to a channel. We may want to consider bookkeeping at
+    // the endpoint level in the future
+    private final LoadingCache<LimitedChannel, Reservoir> costPerChannel =
+            Caffeine.newBuilder().maximumSize(1000).build(upstream -> new ExponentiallyDecayingReservoir());
```

Comment:
you gotta wire clock into these - you can get one from simulation.codahaleClock()